frontend/src/metabase-lib/Question.ts

```diff
@@ -99,16 +101,67 @@ export type QuestionCreatorOpts = {
   dataset_query?: DatasetQuery;
 };
 
+/**
+ * An "opaque type": this is a strong type the TS compiler understands, but you can only create one in this file.
+ *
+ * This technique gives us a way to pass around opaque CLJS values that TS will track for us, and in other files it
+ * gets treated like `unknown` so it can't be examined, manipulated or a new one created.
+ */
+type CljsCard = unknown & { _opaque: typeof CljsCard };
+declare const CljsCard: unique symbol;
+
+function massageForRoundTripCheck(x) {
+  if (typeof x === "undefined" || x == null) {
+    return null;
+  }
+  if (typeof x !== "object") {
+    return x;
+  }
+
+  if (Array.isArray(x)) {
+    const ret = x.map(massageForRoundTripCheck);
+    if (ret.length === 2 && (ret[0] === "field" || ret[0] === "expression")) {
+      ret.push(null);
+    }
+    return ret;
+  }
+
+  const ret = {};
+  for (const k of Object.keys(x)) {
+    ret[k] = massageForRoundTripCheck(x[k]);
+  }
+  return ret;
+}
+
+function diffKeys(a, b) {
+  const keys = [];
+  for (const k of Object.keys(a)) {
+    if (!_.isEqual(a[k], b[k])) {
+      keys.push(k);
+    }
+  }
+  return keys;
+}
```

Comment:
Think we can use [`_.difference`](https://underscorejs.org/#difference) instead?